cartoframes/dataset.py

```diff
@@ -265,24 +275,28 @@ def _get_read_query(self, table_columns, limit=None):
         return query
 
     def get_table_columns(self):
-        """Get column names and types from a table"""
-        query = '''
-            SELECT column_name, data_type
-            FROM information_schema.columns
-            WHERE table_name = '{table}' AND table_schema = '{schema}'
-        '''.format(table=self.table_name, schema=self.schema)
+        """Get column names and types from a table or query result"""
+        if self.query is not None:
+            query = 'SELECT * FROM ({}) _q limit 0'.format(self.query)
+            return get_columns(self.cc, query)
+        else:
+            query = '''
+                SELECT column_name, data_type
+                FROM information_schema.columns
+                WHERE table_name = '{table}' AND table_schema = '{schema}'
+            '''.format(table=self.table_name, schema=self.schema)
 
-        try:
-            table_info = self.cc.sql_client.send(query)
-            return [Column(c['column_name'], pgtype=c['data_type']) for c in table_info['rows']]
-        except CartoException as e:
-            # this may happen when using the default_public API key
-            if str(e) == 'Access denied':
-                query = '''
-                    SELECT *
-                    FROM "{schema}"."{table}" LIMIT 0
-                '''.format(table=self.table_name, schema=self.schema)
-                return get_columns(self.cc, query)
+            try:
+                table_info = self.cc.sql_client.send(query)
+                return [Column(c['column_name'], pgtype=c['data_type']) for c in table_info['rows']]
+            except CartoException as e:
+                # this may happen when using the default_public API key
+                if str(e) == 'Access denied':
```

Comment:
When the `context` contains the `default_public` api_key and you try to get columns from a query with private tables, will the flow end up here?

If so, we should raise a proper `CartoException` to the user.

Reply:
I agree, but we are not handling these type of errors here, for example when the user read o writes a table. I think, `carto-python` should be the responsible

Reply:
OK, I realized I missed the return in line 281 🤐